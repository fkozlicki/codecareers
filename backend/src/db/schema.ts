import { relations, sql } from 'drizzle-orm';
import {
	integer,
	pgEnum,
	pgTable,
	primaryKey,
	text,
	timestamp,
	uuid,
} from 'drizzle-orm/pg-core';

export const users = pgTable('user', {
	id: uuid('id').primaryKey().defaultRandom(),
	firstName: text('first_name'),
	lastName: text('last_name'),
	githubId: integer('github_id'),
	username: text('username'),
	email: text('email').unique(),
	password: text('password'),
});

export const usersRelations = relations(users, ({ many }) => ({
	companies: many(companies),
	applications: many(applications),
}));

export type User = typeof users.$inferSelect;

export const sessions = pgTable('session', {
	id: text('id').primaryKey(),
	userId: uuid('user_id')
		.notNull()
		.references(() => users.id),
	expiresAt: timestamp('expires_at', {
		withTimezone: true,
		mode: 'date',
	}).notNull(),
});

export const companies = pgTable('company', {
	id: uuid('id').primaryKey().defaultRandom(),
	name: text('name').notNull(),
	phoneNumber: text('phone_number').notNull(),
	ownerId: uuid('owner_id').notNull(),
	description: text('description'),
	avatarUrl: text('avatar_url'),
	backgroundUrl: text('background_url'),
});

export const companiesRelations = relations(companies, ({ one, many }) => ({
	owner: one(users, {
		fields: [companies.ownerId],
		references: [users.id],
	}),
	jobOffers: many(jobOffers),
}));

export const levelEnum = pgEnum('level', ['junior', 'mid', 'senior']);
export const employmentTypeEnum = pgEnum('employment_type', [
	'b2b',
	'permanent',
	'mandate',
	'internship',
	'task',
]);
export const workTypeEnum = pgEnum('work_type', [
	'full_time',
	'part_time',
	'internship',
	'freelance',
]);
export const currency = pgEnum('currency', ['pln', 'gbp', 'eur', 'usd']);

export const technologies = pgTable('technology', {
	id: uuid('id').primaryKey(),
	name: text('name').notNull(),
});

export const technologiesRelations = relations(technologies, ({ many }) => ({
	jobOfferTechnologies: many(jobOfferTechnologies),
}));

export const skills = pgTable('skill', {
	id: uuid('id').primaryKey().defaultRandom(),
	name: text('name').notNull(),
});

export const skillsRelations = relations(skills, ({ many }) => ({
	jobOfferSkills: many(jobOfferSkills),
}));

export const jobOffers = pgTable('job_offer', {
	id: uuid('id').primaryKey().defaultRandom(),
	position: text('position').notNull(),
	description: text('description').notNull(),
	level: levelEnum('level').notNull(),
	employmentType: employmentTypeEnum('employment_type').notNull(),
	workType: workTypeEnum('work_type').notNull(),
	salaryFrom: integer('salary_from').notNull(),
	salaryTo: integer('salary_to').notNull(),
	salaryCurrency: currency('currency').notNull(),
	companyId: uuid('company_id').notNull(),
	createdAt: timestamp('created_at', {
		withTimezone: true,
		mode: 'date',
	})
		.notNull()
		.defaultNow(),
});

export const jobOffersRelations = relations(jobOffers, ({ many, one }) => ({
	jobOfferSkills: many(jobOfferSkills),
	jobOfferTechnologies: many(jobOfferTechnologies),
	applications: many(applications),
	company: one(companies, {
		fields: [jobOffers.companyId],
		references: [companies.id],
	}),
}));

export const jobOfferTechnologies = pgTable(
	'job_offer_technology',
	{
		jobOfferId: uuid('job_offer_id')
			.notNull()
			.references(() => jobOffers.id),
		technologyId: uuid('technology_id')
			.notNull()
			.references(() => technologies.id),
	},
	(t) => ({
		pk: primaryKey({ columns: [t.jobOfferId, t.technologyId] }),
	})
);

export const jobOfferSkills = pgTable(
	'job_offer_skill',
	{
		jobOfferId: uuid('job_offer_id')
			.notNull()
			.references(() => jobOffers.id),
		skillId: uuid('skill_id')
			.notNull()
			.references(() => skills.id),
	},
	(t) => ({
		pk: primaryKey({ columns: [t.jobOfferId, t.skillId] }),
	})
);

export const applications = pgTable('application', {
	id: uuid('id').primaryKey().defaultRandom(),
	createdAt: timestamp('created_at', {
		withTimezone: true,
		mode: 'date',
	})
		.notNull()
		.defaultNow(),
	userId: uuid('user_id').notNull(),
	jobOfferId: uuid('job_offer_id').notNull(),
	cv: text('cv'),
	introduction: text('introduction'),
});

export const applicationsRelations = relations(applications, ({ one }) => ({
	jobOffer: one(jobOffers, {
		fields: [applications.jobOfferId],
		references: [jobOffers.id],
	}),
	user: one(users, {
		fields: [applications.userId],
		references: [users.id],
	}),
}));
