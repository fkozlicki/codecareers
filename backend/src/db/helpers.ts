import { sql } from 'drizzle-orm';
import { db } from '.';

export const clearDb = async () => {
	const query = sql<string>`SELECT table_name 
    FROM information_schema.tables 
    WHERE table_schema = 'public' 
    AND table_type = 'BASE TABLE'`;

	const tables = await db.execute(query);

	for (let table of tables.rows) {
		const query = sql.raw(
			`TRUNCATE TABLE public."${table.table_name}" CASCADE`
		);
		await db.execute(query);
	}
};
