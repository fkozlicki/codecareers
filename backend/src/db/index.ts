import { drizzle } from 'drizzle-orm/node-postgres';
import pg from 'pg';
import * as schema from './schema.js';
import dotnev from 'dotenv';

dotnev.config({ path: `.env.${process.env.NODE_ENV ?? 'development'}` });

const pool = new pg.Pool({
	connectionString: process.env.DATABASE_URL!,
});

export const db = drizzle(pool, {
	schema,
});
