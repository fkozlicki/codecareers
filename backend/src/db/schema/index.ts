export * from './application';
export * from './chat';
export * from './chatUser';
export * from './company';
export * from './jobOffer';
export * from './jobOfferSkill';
export * from './jobOfferTechnology';
export * from './message';
export * from './recruitment';
export * from './session';
export * from './skill';
export * from './technology';
export * from './user';
