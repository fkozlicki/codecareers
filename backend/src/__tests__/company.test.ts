import { Argon2id } from 'oslo/password';
import supertest from 'supertest';
import { app } from '..';
import { db } from '../db';
import { clearDb } from '../db/helpers';
import {
	companies,
	jobOffers,
	skills,
	technologies,
	users,
} from '../db/schema';

const password = 'S3cReT123';

let adam;
let jon;
let adamSession = '';
let jonSession = '';
let google;
let skillAi;
let technologyPython;
let seniorWebDeveloper;

const companyRequest = {
	name: 'Amazon',
	description: 'Lorem ipsum',
	phoneNumber: '222 222 2222',
};

beforeEach(async () => {
	await clearDb();
});

beforeEach(async () => {
	const hashedPassword = await new Argon2id().hash(password);
	adam = (
		await db
			.insert(users)
			.values({
				firstName: 'Adam',
				lastName: 'Joseph',
				email: 'adamjospeh@gmail.com',
				password: hashedPassword,
			})
			.returning()
	)[0];

	jon = (
		await db
			.insert(users)
			.values({
				firstName: 'Jon',
				lastName: 'Snow',
				email: 'jon.snow@gmail.com',
				password: hashedPassword,
			})
			.returning()
	)[0];

	google = (
		await db
			.insert(companies)
			.values({
				name: 'Google',
				ownerId: adam!.id,
				phoneNumber: '222 222 2222',
			})
			.returning()
	)[0];

	seniorWebDeveloper = (
		await db
			.insert(jobOffers)
			.values({
				companyId: google.id,
				description: '',
				employmentType: 'b2b',
				level: 'senior',
				position: 'Senior Web Developer',
				salaryCurrency: 'pln',
				salaryFrom: 25000,
				salaryTo: 30000,
				workType: 'full_time',
			})
			.returning()
	)[0];

	skillAi = (await db.insert(skills).values({ name: 'AI' }).returning())[0];
	technologyPython = (
		await db.insert(technologies).values({ name: 'Python' }).returning()
	)[0];

	// sign in
	const adamResult = await supertest(app).post('/login/credentials').send({
		email: 'adamjospeh@gmail.com',
		password,
	});
	adamSession = adamResult.header['set-cookie'];

	const jonResult = await supertest(app).post('/login/credentials').send({
		email: 'jon.snow@gmail.com',
		password,
	});

	jonSession = jonResult.header['set-cookie'];
});

describe('Company service', () => {
	describe('[POST] /companies', () => {
		it('Create company, when user is signed in and body is valid', async () => {
			const { statusCode, body } = await supertest(app)
				.post('/companies')
				.set('cookie', adamSession)
				.send(companyRequest);

			expect(statusCode).toBe(201);
			expect(body).toMatchObject({
				company: {
					name: 'Amazon',
					description: 'Lorem ipsum',
					phoneNumber: '222 222 2222',
					avatarUrl: null,
					backgroundUrl: null,
				},
			});
		});

		it('Throw 401, when user is not signed in', async () => {
			const { statusCode, body } = await supertest(app)
				.post('/companies')
				.send(companyRequest);

			expect(statusCode).toBe(401);
			expect(body).toEqual({ message: 'Auth required' });
		});

		it('Throw 400, when credentials are invalid', async () => {
			const invalidCredentials = [
				// missing name
				{ description: 'Lorem ipsum', phoneNumber: '222 222 2222' },
				// missing description
				{ name: 'Amazon', phoneNumber: '222 222 2222' },
				// missing phone number
				{ name: 'Amazon', description: 'Lorem ipsum' },
			];

			for (const credentials of invalidCredentials) {
				const { statusCode, body } = await supertest(app)
					.post('/companies')
					.set('cookie', adamSession)
					.send(credentials);

				expect(statusCode).toBe(400);
				expect(body).toEqual({ error: 'Invalid data' });
			}
		});
	});

	describe('[GET] /companies/:id', () => {
		it('Return company, when user have access to it', async () => {
			const { statusCode, body } = await supertest(app)
				.get(`/companies/${google!.id}`)
				.set('cookie', adamSession);

			expect(statusCode).toBe(200);
			expect(body.company).toMatchSnapshot({
				id: expect.any(String),
			});
		});

		it('Throw 401, when user is not signed in', async () => {
			const { statusCode, body } = await supertest(app).get(
				`/companies/${google!.id}`
			);

			expect(statusCode).toBe(401);
			expect(body).toEqual({ message: 'Auth required' });
		});

		it('Throw 403, when user does not have access to company', async () => {
			const { statusCode, body } = await supertest(app)
				.get(`/companies/${google!.id}`)
				.set('cookie', jonSession);

			expect(statusCode).toBe(403);
			expect(body).toEqual({
				message: "You don't have permission to access this data",
			});
		});
	});

	describe('[GET] /companies', () => {
		it("Return list of Adam's companies", async () => {
			const { statusCode, body } = await supertest(app)
				.get(`/companies/?userId=${adam!.id}`)
				.set('cookie', adamSession);

			expect(statusCode).toBe(200);
			for (const company of body.companies) {
				expect(company).toMatchSnapshot({ id: expect.any(String) });
			}
		});
	});

	describe('[PUT] /companies/:id', () => {
		it('Should update company', async () => {
			const { statusCode, body } = await supertest(app)
				.put(`/companies/${google!.id}`)
				.set('cookie', adamSession)
				.send({
					name: 'Google Inc.',
					description: 'This is Google Inc. after update',
				});

			expect(statusCode).toBe(200);
			expect(body.company).toMatchSnapshot({ id: expect.any(String) });
		});

		it('Throw 401, when user is not signed in', async () => {
			const { statusCode, body } = await supertest(app)
				.put(`/companies/${google!.id}`)
				.send({
					name: 'Google Inc.',
					description: 'This is Google Inc. after update',
				});

			expect(statusCode).toBe(401);
			expect(body).toEqual({ message: 'Auth required' });
		});

		it('Throw 403, when user does not have access to company', async () => {
			const { statusCode, body } = await supertest(app)
				.put(`/companies/${google!.id}`)
				.set('cookie', jonSession)
				.send({
					name: 'Google Inc.',
					description: 'This is Google Inc. after update',
				});

			expect(statusCode).toBe(403);
			expect(body).toEqual({
				message: "You don't have permission to access this data",
			});
		});
	});

	describe('[POST] /comapnies/:id/job-offers', () => {
		it('Should create new job offer', async () => {
			const { statusCode, body } = await supertest(app)
				.post(`/companies/${google!.id}/job-offers`)
				.set('cookie', adamSession)
				.send({
					position: 'Junior AI Engeneer',
					description: 'We are looking for talented AI Engeneer',
					level: 'junior',
					employmentType: 'internship',
					workType: 'part_time',
					salaryFrom: 6000,
					salaryTo: 8000,
					salaryCurrency: 'pln',
					skills: [{ label: 'AI', value: skillAi!.id }],
					technologies: [{ label: 'Python', value: technologyPython!.id }],
				});

			expect(statusCode).toBe(201);
			expect(body.jobOffer).toMatchSnapshot({
				id: expect.any(String),
				createdAt: expect.any(String),
			});
		});
	});

	describe('[GET] /comapnies/:id/job-offers', () => {
		it('Should return company job offers', async () => {
			const { statusCode, body } = await supertest(app)
				.get(`/companies/${google!.id}/job-offers`)
				.set('cookie', adamSession);

			expect(statusCode).toBe(200);
			for (const jobOffer of body.jobOffers) {
				expect(jobOffer).toMatchSnapshot({
					id: expect.any(String),
					createdAt: expect.any(String),
				});
			}
		});
	});
});
