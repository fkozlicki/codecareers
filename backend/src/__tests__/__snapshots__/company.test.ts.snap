// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Company service [GET] /comapnies/:id/job-offers Should return company job offers 1`] = `
{
  "company": {
    "avatarUrl": null,
    "backgroundUrl": null,
    "description": null,
    "id": Any<String>,
    "name": "Google",
    "ownerId": Any<String>,
    "phoneNumber": "222 222 2222",
  },
  "companyId": Any<String>,
  "createdAt": Any<String>,
  "description": "",
  "employmentType": "b2b",
  "id": Any<String>,
  "level": "senior",
  "position": "Senior Web Developer",
  "published": false,
  "salaryCurrency": "pln",
  "salaryFrom": 25000,
  "salaryTo": 30000,
  "workType": "full_time",
}
`;

exports[`Company service [GET] /companies Return list of Adam's companies 1`] = `
{
  "avatarUrl": null,
  "backgroundUrl": null,
  "description": null,
  "id": Any<String>,
  "name": "Google",
  "phoneNumber": "222 222 2222",
}
`;

exports[`Company service [GET] /companies/:id Return company, when user have access to it 1`] = `
{
  "avatarUrl": null,
  "backgroundUrl": null,
  "description": null,
  "id": Any<String>,
  "name": "Google",
  "phoneNumber": "222 222 2222",
}
`;

exports[`Company service [POST] /comapnies/:id/job-offers Should create new job offer 1`] = `
{
  "companyId": Any<String>,
  "createdAt": Any<String>,
  "description": "We are looking for talented AI Engeneer",
  "employmentType": "internship",
  "id": Any<String>,
  "level": "junior",
  "position": "Junior AI Engeneer",
  "published": false,
  "salaryCurrency": "pln",
  "salaryFrom": 6000,
  "salaryTo": 8000,
  "workType": "part_time",
}
`;

exports[`Company service [PUT] /companies/:id Should update company 1`] = `
{
  "avatarUrl": null,
  "backgroundUrl": null,
  "description": "This is Google Inc. after update",
  "id": Any<String>,
  "name": "Google Inc.",
  "phoneNumber": "222 222 2222",
}
`;
